presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java

```diff
@@ -0,0 +1,235 @@
+        if (matches.size() > 1) {
+            throwAmbiguousAttributeException(node, name);
+        }
+
+        if (matches.isEmpty()) {
+            if (isColumnReference(name, relation)) {
+                return Optional.empty();
+            }
+            Scope boundary = this;
+            while (!boundary.queryBoundary) {
+                if (boundary.parent.isPresent()) {
+                    boundary = boundary.parent.get();
+                }
+                else {
+                    return Optional.empty();
+                }
+            }
+            if (boundary.parent.isPresent()) {
+                // jump over the query boundary
+                return boundary.parent.get().resolveField(node, name, false);
+            }
+            return Optional.empty();
+        }
+        else {
+            return Optional.of(asResolvedField(getOnlyElement(matches), local));
+        }
+    }
+
+    private ResolvedField asResolvedField(Field field, boolean local)
+    {
+        int fieldIndex = relation.indexOf(field);
+        return new ResolvedField(this, field, fieldIndex, local);
+    }
+
+    public boolean isColumnReference(QualifiedName name)
+    {
+        Scope current = this;
+        while (current != null) {
+            if (isColumnReference(name, current.relation)) {
+                return true;
+            }
+
+            current = current.parent.orElse(null);
+        }
+
+        return false;
+    }
+
+    private boolean isColumnReference(QualifiedName name, RelationType relation)
+    {
+        while (name.getPrefix().isPresent()) {
+            name = name.getPrefix().get();
+            if (!relation.resolveFields(name).isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public Optional<WithQuery> getNamedQuery(String name)
+    {
+        if (namedQueries.containsKey(name)) {
+            return Optional.of(namedQueries.get(name));
+        }
+
+        if (parent.isPresent()) {
+            return parent.get().getNamedQuery(name);
+        }
+
+        return Optional.empty();
+    }
+
+    public boolean isApproximate()
+    {
+        return approximate;
+    }
+
+    public static final class Builder
+    {
+        private RelationType relationType = new RelationType();
+        private Optional<Boolean> approximate = Optional.empty();
+        private boolean queryBoundary = false;
+        private final Map<String, WithQuery> namedQueries = new HashMap<>();
+        private Optional<Scope> parent = Optional.empty();
+
+        public Builder withRelationType(RelationType relationType)
+        {
+            this.relationType = requireNonNull(relationType, "relationType is null");
+            return this;
+        }
+
+        public Builder withParent(Scope parent)
+        {
+            this.parent = Optional.of(parent);
+            return this;
+        }
+
+        public Builder withQueryBoundary(boolean queryBoundary)
```

Comment:
Is this ever called with `false`? I don't think there's any use case for that.

I'd change this method to take no arguments and rename it to `markQueryBoundary` or `makeQueryBoundary`, or maybe even have a different `builder()` method in `Scope` to create a nested scope.